[Feat] 챌린지 프로필 인증현황 API 구현

yangjiae12 opened the pull request:
## #️⃣ 연관된 이슈

> 관련된 이슈 번호를 적어주세요.\
> Close #90 

## ✨ 작업 내용 (Summary)

> 이번 PR에서 작업한 내용을 간략히 설명해주세요. (이미지 첨부 가능)

이번 PR에서는 챌린지 인증 피드 조회와 인증 인원 통계 조회 API를 구현했습니다. 클라이언트의 요청 편의성을 높이기 위해 RoundNumber를 기반으로 조회하도록 설계하였습니다.

#### 1. 인증 피드 조회 API 구현 (GET /api/v1/verifications)
- 챌린지 ID와 라운드 번호(1, 2, 3...)를 통해 해당 라운드의 인증 글 목록을 조회합니다.
- 정렬 로직
  - 해결되지 않은 질문 (최상단)
  - 그 외 모든 글 (최신순 정렬)
- 챌린지의 verificationType에 따라 자동으로 필터링 하여 반환합니다.

#### 2. 인증 인원 통계 API 구현 (GET /api/v1/verifications/stat)
- 현재 진행 중인 라운드의 실시간 인증 인원 수와 전체 참가자 수를 반환합니다.
- 유효성 검증 추가: 챌린지가 진행 중(ONGOING)이거나 모집 중(RECRUITING)이 아닌 경우 예외(CHALLENGE_NOT_IN_PROGRESS)를 발생시켜 불필요한 조회를 차단했습니다.

---

## ✅ 변경 사항 체크리스트

> 다음 항목들을 확인하고 체크해주세요.

- [ ] 코드에 영향이 있는 모든 부분에 대한 테스트를 작성하고 실행했나요?
- [ ] 문서를 작성하거나 수정했나요? (필요한 경우)
- [ ] 중요한 변경 사항이 팀에 공유되었나요?

---

## 🧪 테스트 결과

> 코드 변경에 대해 테스트를 수행한 결과를 요약해주세요.

* **테스트 환경:** 로컬
* **테스트 방법:** 스웨거
* **결과 요약:** 테스트 통과 및 예외 처리 확인 완료

---

## 📸 스크린샷

> 관련된 스크린샷 또는 GIF가 있다면 여기에 첨부해주세요.

| 기능 | 스크린샷 |
|---|--------------|
| 챌린지 인증 현황 피드 조회 | <img width="1175" height="746" alt="image" src="https://github.com/user-attachments/assets/0aa284e3-ad3e-4ebc-9c23-534ccc8de126" /> |
| 챌린지 인증 현황 통계 조회 | <img width="1175" height="412" alt="image" src="https://github.com/user-attachments/assets/4d3d7e33-8d80-4bbb-bf17-12a7b787fe77" /> |


---

## 💬 리뷰 요구사항

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요.
- VerificationRepository의 JPQL 쿼리: 미해결 질문을 최상단으로 올리는 CASE WHEN 정렬 로직이 성능이나 가독성 면에서 적절한지 확인 부탁드립니다.

---

## 📎 참고 자료

> 관련 문서, 레퍼런스 링크 등이 있다면 여기에 첨부해주세요.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새로운 기능**
  * 인증 피드 조회 및 페이징 지원 추가
  * 인증 통계(헤더) 조회 기능 추가
  * 내 인증 프로필(내 피드·통계) 조회 추가
  * 인증 해결 표시 및 댓글 채택 표시 기능 추가

* **개선사항**
  * API 경로 통일 (/verifications) 및 요청 검증 강화
  * 응답 DTO 확장으로 피드·통계·생성 응답 개선

* **데이터베이스**
  * 인증·댓글에 기본 상태 필드(is_resolved, is_adopted) 및 댓글 내용 NOT NULL 적용

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yangjiae12 committed:
feat: 인증 피드 조회를 위한 도메인 필드 및 레포지토리 쿼리 구현 #90

- Verification(해결 여부), Comment(채택 여부) 엔티티 필드 추가
- 미해결 질문 우선 정렬 및 상태 필터링을 위한 JPQL 쿼리 구현 (VerificationRepository)
- 챌린지 ID와 라운드 번호 기반 조회 메서드 추가 (RoundRepository)
- 관련 ErrorCode 추가

yangjiae12 committed:
feat: 챌린지 인증 피드 조회 API 구현 #90

- 인증 피드 조회 비즈니스 로직 구현 (Service)
  - 챌린지 및 라운드 유효성 검증
  - VerificationType(PHOTO/TEXT)에 따른 필터링 로직
- DTO 및 Entity-DTO 변환 컨버터 구현
- 인증 피드 조회 엔드포인트 구현 (GET /api/v1/verifications)

yangjiae12 committed:
feat: 챌린지 인증 인원 통계 조회 API 구현 #90

- 인증 인원 통계 조회 엔드포인트 구현 (GET /api/v1/verifications/stat)
- Verification 도메인의 Controller, Service, DTO, Converter 계층 구현
- 챌린지 상태(ONGOING, RECRUITING) 검증 및 예외 처리 로직 추가
- 현재 라운드 기준 실시간/최근 인증 인원 수 및 전체 참가자 수 집계 로직 구현

yangjiae12 committed:
chore: 챌린지 프로필 인증 현황 API 엔드포인트 수정 #90

yangjiae12 committed:
Merge branch 'develop' into feat/90-verification-list

yangjiae12 committed:
refactor: 인증 피드 응답 DTO의 링크 필드 변경 (linkUrl -> hasLink) #90

- 프론트엔드 UI 요구사항(아이콘 표시)에 맞춰 전체 URL 문자열 대신 존재 여부(boolean)만 반환하도록 변경
- VerificationConverter에 textUrl 유효성 검사 로직 추가

yangjiae12 committed:
fix: 인증 통계 조회 시 currentRound NPE 방지 #90

- 데이터 정합성 오류 등으로 인해 currentRound가 null일 경우 500 에러 대신 0명 반환 처리

yangjiae12 committed:
refactor: 인증 집계 쿼리 반환 타입 불일치 수정 (Integer -> Long) #90

- Repository: JPQL COUNT 쿼리 스펙에 맞춰 반환 타입을 Long으로 변경
- Service: Long 타입을 받아 int로 변환하여 처리하도록 로직 수정

yangjiae12 committed:
Merge branch 'feat/90-verification-list' of https://github.com/Hrr-HabbitRoutineResult/hrr-server into feat/90-verification-list

yangjiae12 committed:
fix: Verification 엔티티 isResolved 필드에 @Builder.Default 추가 #90

- Lombok @Builder 사용 시 필드 초기화(= false)가 무시되는 문제 해결
- 생성 시 null이 들어가는 것을 방지하여 DB Not Null 제약조건 준수

yangjiae12 committed:
chore: StatDto.baseDate 예시 포맷이 실제 JsonFormat과 통일 #90

yangjiae12 committed:
Merge branch 'develop' into feat/90-verification-list

yangjiae12 committed:
feat: Verification 및 Comment 테이블 관련 마이그레이션 스크립트 작성 (V2.8)

- V2.8__alter_verification_and_comment.sql 추가
- Verification 테이블: is_resolved 컬럼 추가
- Comment 테이블: content, is_adopted, verification_id 컬럼 및 FK 추가

yangjiae12 committed:
Merge branch 'develop' into feat/90-verification-list

yangjiae12 committed:
fix: DTO 구조 변경에 따른 Controller/Service 컴파일 에러 수정 #90

- Controller/Service의 반환 타입을 구체적인 내부 클래스(CreateResponseDto 등)로 명시하여 타입 불일치 해결
- VerificationConverter의 빌더 패턴 사용 방식 수정 (Wrapper Class 적용)
- 인증 생성(Create) 및 조회(Read) 비즈니스 로직 통합 완료

yangjiae12 committed:
fix: Flyway 마이그레이션 버전 중복 및 구문 파싱 오류 해결 #90

- develop 브랜치와의 버전 충돌 해결을 위해 스크립트 버전을 V2.8 -> V2.12로 변경
- Stored Procedure의 DELIMITER 및 구문 오류 수정하여 멱등성 확보
- 인증 및 댓글 테이블 컬럼 추가 스크립트 정리

yc3697 commented on src/main/java/com/hrr/backend/domain/verification/repository/VerificationRepository.java:
```diff
@@ -29,6 +30,49 @@ boolean existsTodayVerification(
+            "AND v.type = :type " +
+            "AND v.status = :status " +
+            "ORDER BY " +
+            "  CASE WHEN (v.isQuestion = true AND v.isResolved = false) THEN 0 ELSE 1 END ASC, " +
```
이렇게도 할 수 있구나,,, 가독성 굿입니다요

yc3697 commented on src/main/java/com/hrr/backend/domain/round/repository/RoundRecordRepository.java:
```diff
@@ -42,4 +42,12 @@ Optional<RoundRecord> findByUserChallengeAndRoundId(
+
+    // 내 총 인증 횟수 합계
+    @Query("SELECT COALESCE(SUM(r.verificationCount), 0) FROM RoundRecord r WHERE r.userChallenge.id = :userChallengeId")
+    Integer sumVerificationCountByUserChallengeId(@Param("userChallengeId") Long userChallengeId);
```
리뷰 해결해주세영~

yc3697 commented on src/main/java/com/hrr/backend/domain/verification/controller/VerificationController.java:
```diff
@@ -37,8 +65,7 @@ public ApiResponse<VerificationResponseDto> createTextVerification(
     ) {
         Long userId = userDetails.getUser().getId();
 
-        // roundId는 프론트에서 받지 않고, Service에서 현재 라운드를 자동 조회
```
프론트 배려가 꼼꼼하니 야무집니다

yc3697 commented on src/main/resources/db/migration/V2.12__alter_verification_and_comment.sql:
```diff
@@ -0,0 +1,33 @@
+ALTER TABLE verification ADD COLUMN is_resolved BIT(1) DEFAULT 0 NOT NULL;
+END IF;
+
+    -- 2. Comment 테이블: is_adopted 컬럼 추가 (존재하지 않을 경우에만)
```
content가 not null로 바뀐 것도 반영해주세요!

yc3697 reviewed (changes requested):
스크립트 수정이랑 Integer->Long만 변경해주세요~

yangjiae12 committed:
fix: RoundRecordRepository 집계 함수 반환 타입 Long으로 변경 #90

- JPQL의 COUNT, SUM 함수 반환값이 Long임에 따라 Repository 메서드 반환 타입 수정
- Service 및 DTO에서 불필요한 형변환 제거하고 Long 타입으로 통일

yangjiae12 committed:
fix: Comment 엔티티와 DB 스키마 불일치 해결 #90

- content 컬럼의 Nullable 설정을 Entity(nullable=false)에 맞춰 수정

Hminkyung reviewed (approved):
굿굿 좋아요 제 로직이랑은 별다른 문제 없는것 같습니다

yc3697 reviewed (approved):
LGTM~
(한 번 써보고 싶었어요 ㅎㅎ)
